crates/red_knot_python_semantic/src/types.rs

```diff
@@ -3965,6 +3972,86 @@ impl<'db> Class<'db> {
         symbol(db, scope, name)
     }
 
+    /// Returns the `name` attribute of an instance of this class.
+    ///
+    /// The attribute could be defined in the class body, but it could also be an implicitly
+    /// defined attribute that is only present in a method (typically `__init__`).
+    ///
+    /// The attribute might also be defined in a superclass of this class.
+    pub(crate) fn instance_member(self, db: &'db dyn Db, name: &str) -> Symbol<'db> {
+        for superclass in self.iter_mro(db) {
+            match superclass {
+                ClassBase::Dynamic(_) => {
+                    return todo_type!("instance attribute on class with dynamic base").into();
+                }
+                ClassBase::Class(class) => {
+                    let member = class.own_instance_member(db, name);
+                    if !member.is_unbound() {
+                        return member;
+                    }
+                }
+            }
+        }
+
+        // TODO: The symbol is not present in any class body, but it could be implicitly
+        // defined in `__init__` or other methods anywhere in the MRO.
+        todo_type!("implicit instance attribute").into()
+    }
+
+    /// A helper function for `instance_member` that looks up the `name` attribute only on
+    /// this class, not on its superclasses.
+    fn own_instance_member(self, db: &'db dyn Db, name: &str) -> Symbol<'db> {
+        // TODO: There are many things that are not yet implemented here:
+        // - `typing.ClassVar` and `typing.Final`
+        // - Proper diagnostics
+        // - Handling of possibly-undeclared/possibly-unbound attributes
+        // - The descriptor protocol
+
+        let body_scope = self.body_scope(db);
+        let table = symbol_table(db, body_scope);
+
+        if let Some(symbol) = table.symbol_id_by_name(name) {
+            let use_def = use_def_map(db, body_scope);
+
+            let declarations = use_def.public_declarations(symbol);
+
+            match declarations_ty(db, declarations) {
+                Ok(Symbol::Type(declared_ty, _)) => {
+                    if let Some(function) = declared_ty.into_function_literal() {
+                        // TODO: Eventually, we are going to process all decorators correctly. This is
+                        // just a temporary heuristic to provide a broad categorization into properties
+                        // and non-property methods.
+                        if function.has_decorator(db, KnownClass::Property.to_class_literal(db)) {
+                            todo_type!("@property").into()
+                        } else {
+                            todo_type!("bound method").into()
+                        }
```

Comment:
I'm really hopeful that we can keep our special-casing of properties to a minimum if we have a solid implementation of the descriptor protocol, since they really are just descriptor instances at the end of the day. But this seems fine for now, since properties are by far the most common descriptors around, and this improves our TODO messages, making it clear why we infer bad types for various `@property` attributes for the time being